src/main/java/htsjdk/samtools/cram/build/Sam2CramRecordFactory.java

```diff
@@ -291,36 +271,62 @@ private void addInsertion(final List<ReadFeature> features, final int zeroBasedP
         }
     }
 
-    private void addSubstitutionsAndMaskedBases(final CramCompressionRecord cramRecord, final List<ReadFeature> features, final int fromPosInRead, final int
+    /**
+     * Processes a stretch of read bases marked as match or mismatch and emits appropriate read features.
+     * Briefly the algorithm is:
+     * <ul><li>emit nothing for a read base matching corresponding reference base.</li>
+     * <li>emit a {@link Substitution} read feature for each ACTGN-ACTGN mismatch.</li>
+     * <li>emit {@link ReadBase} for a non-ACTGN mismatch. The side effect is the quality score stored twice.</li>
+     * <p>
+     * The rest of the method handles quality score read features added if refSNPs is specified.
+     * <p>
+     * IMPORTANT: reference and read bases are always compared for match/mismatch in upper case due to BAM limitations.
+     *
+     * @param cramRecord           a record to work on
+     * @param features             a list of read features to add to
+     * @param fromPosInRead        a zero based position in the read to start with
+     * @param alignmentStartOffset offset into the reference array
+     * @param nofReadBases         how many read bases to process
+     * @param bases                the read bases array
+     * @param qualityScore         the quality score array
+     */
+    void addSubstitutionsAndMaskedBases(final CramCompressionRecord cramRecord, final List<ReadFeature> features, final int fromPosInRead, final int
             alignmentStartOffset, final int nofReadBases, final byte[] bases, final byte[] qualityScore) {
         int oneBasedPositionInRead;
         final boolean noQS = (qualityScore.length == 0);
 
-        int i;
         boolean qualityAdded;
         byte refBase;
-        for (i = 0; i < nofReadBases; i++) {
+        for (int i = 0; i < nofReadBases; i++) {
             oneBasedPositionInRead = i + fromPosInRead + 1;
             final int referenceCoordinates = cramRecord.alignmentStart + i + alignmentStartOffset - 1;
```

Comment:
It would be a lot cleaner  if oneBasedPositionInRead and referenceCoordinates had their initial value calculated once, and then were just incremented in the for loop's increment expression along with nofReadBases.